Update package infrastructure

pratikunterwegs opened the pull request:
This PR adds some basic R package infrastructure:

1. Updates or adds GHA workflows:
  - Updates the R CMD check and test coverage workflows using {usethis} autogenerated YAMLs
  - Adds the `render-readme` workflow to follow the Rmd -> md rendering way of working
  - Adds linting, citation file generation, and license year updating workflows
2. Adds some basic information to the Readme.Rmd, adds a dummy vignette to populate the 'Get started' page, and adds a reference library as a JSON.
3. Adds the MIT license text as autogenerated from {usethis}
4. Adds Readme badges
5. Adds an R project file to make it easier to work with in RStudio
6. Updates `.Rbuildignore`, `tests/testthat.R`, and `NAMESPACE`
7. Adds package documentation using `usethis::use_package_doc()`
8. Adds basic website structure in `_pkgdown.yml`
9. Updates `DESCRIPTION` with preliminary package title and details.
10. Adds spellchecking and a WORDLIST for known words.

richfitz reviewed (changes requested):
I would strongly prefer actions that do not cause changes to the branch that is being modified, I believe that is happening in two places.

Other minor comments, suggestions, and nudges toward how most of our repos are configured included inline

pratikunterwegs commented on .github/workflows/render_readme.yaml:
```diff
@@ -0,0 +1,97 @@
+          )
+        shell: Rscript {0}
+
+      - name: Commit files
```
The current branch - https://github.com/j-idea/daedalus/pull/1/commits/1275793f4fb0d93ea6f47e3f533031cba80a5e66 is an example.

pratikunterwegs commented on .lintr:
```diff
@@ -1,7 +1,24 @@
-    cyclocomp_linter = NULL
+linters: all_linters(
+    packages = c("lintr"),
+    object_name_linter = NULL,
```
I've previously seen complaints on long variable names, which can be common if we're aiming for descriptive naming, which IMO is easier to understand. I _think_ it'll also complain on single letter names which folks might want to use/see used in ODE systems.

pratikunterwegs commented on .lintr:
```diff
@@ -1,7 +1,24 @@
-    todo_comment_linter = NULL,
-    cyclocomp_linter = NULL
+linters: all_linters(
+    packages = c("lintr"),
```
I'll remove it - can re-add if we choose to include linters from another package.

pratikunterwegs replied:
Removed in https://github.com/j-idea/daedalus/pull/1/commits/901a5ad480834149bad03e42a16ef22f64658e78

pratikunterwegs commented on README.Rmd:
````diff
@@ -0,0 +1,47 @@
+)
+```
+
+# {{ packagename }}: Optimise economic, social and health trade-offs in a pandemic
````
The `render-readme` workflow will replace all instances of `{{ packagename }}` with the current package name. Sure, it probably won't change but it might be good to have flexibility. This is taken from a template repo so it's set up to be package name agnostic.

richfitz replied:
really, if the package name changes, this will be the least of the effort to fix

david-mears-2 replied:
@richfitz You've misspelt 'daedalus' in your suggested change. Which is, in itself, a fantastic argument for templating rather than having a human type out the package name.

richfitz replied:
or an  indictment of the package name 🙃 

The package name will appear in many places in the package eventually, I'll get better at spelling it over time I hope, though my track record with some words is poor...

pratikunterwegs commented on README.Rmd:
````diff
@@ -0,0 +1,47 @@
+
+You can install the development version of daedalus from [GitHub](https://github.com/) using the _pak_ package.
+
+``` r
````
1. R-universe - yes, will add that once it's up. I think a J-IDEA one would be a good idea (name pending).
2. {pak} vs {remotes}: I haven't had an issue with {pak} in the past and I think it's nicer for end users who this is aimed at (better dependency display, cached downloads). I didn't encounter an issues with installing `mrc-ide/dust2` using {pak}, does this only error on Windows?

richfitz replied:
It's more things for users to install, and the interface to pak has never properly stabilised. Please use remotes if you need a direct install from source.  The no-upgrades is because the Rstudio folks have never made that work reliably on windows where filelocking is more aggressive than people think it is, and it will enthusiastically update dependencies, leaving corrupt libraries in its wake

pratikunterwegs replied:
Done.

pratikunterwegs commented on _pkgdown.yml:
```diff
@@ -0,0 +1,10 @@
+ - title: All functions
+   desc: All package functions
+ - contents:
+   - starts_with("*")
```
There's only the dummy function``%||%`` right now so this will need updating once more functions are added - I don't intend to keep it this way, that would be pretty poor organisation.

pratikunterwegs commented on R/daedalus-package.R:
```diff
@@ -0,0 +1,6 @@
+#' @keywords internal
+"_PACKAGE"
+
+## usethis namespace: start
```
They're autogenerated from [`usethis::use_package_doc()`](https://usethis.r-lib.org/reference/use_package_doc.html) and bring up the package Readme when users enter `?daedalus`.

pratikunterwegs commented on .github/workflows/update-copyright-year.yml:
```diff
@@ -0,0 +1,37 @@
+name: Update copyright year(s) in license file
```
Since it's on a cron schedule it'll only trigger a run in the new year, so I'm not sure a check on the year would work - it'll always be out of date when the job runs.

pratikunterwegs commented on .lintr:
```diff
@@ -1,7 +1,24 @@
+    undesirable_function_linter(
+      modify_defaults(
+        default_undesirable_functions,
+        library = NULL # too many false positive in too many files
```
Fixed in https://github.com/j-idea/daedalus/pull/1/commits/901a5ad480834149bad03e42a16ef22f64658e78

pratikunterwegs commented:
Thanks for taking a look @richfitz - I've addressed most of your comments.

> I would strongly prefer actions that do not cause changes to the branch that is being modified, I believe that is happening in two places.

I can remove `render-readme` if you'd like but I've never had an issue with it on other repos, is there a specific concern around it? One wrinkle I do know of with auto-generated commits like the one from `render-readme` is that they don't play nice with organisation-wide workflows but we don't have those as it is.

Is there another workflow changing the PR branch - the `update-citation` workflow makes a separate PR.

david-mears-2 commented on NEWS.md:
```diff
@@ -0,0 +1,9 @@
+# daedalus 0.0.1
+
+Initial package setup including GitHub Actions workflows and code quality checks.
+
+- Updated GH Actions workflows for R CMD check (fails on 'NOTE') and code test coverage;
+
+- Added linting, citation update, Readme rendering, and license year update workflows;
+
+- Added spellchecking and a known words list.
```
Is this file for, like, release notes? Is it auto-generated from commit messages?

pratikunterwegs replied:
It can indeed be incorporated into release notes via `usethis::use_github_release()` which is best practice for making a GH release of an R package. It also appears on the {pkgdown} website as "Changelog".

It isn't auto-generated from the commit messages, no - it's manually written. Some packages have quite detailed ones (my preference), while others don't have one at all.
